packages/react/src/components/Listbox/Listbox.tsx

```diff
@@ -0,0 +1,216 @@
+import React, {
+  forwardRef,
+  useCallback,
+  useState,
+  useLayoutEffect,
+  useEffect
+} from 'react';
+import { ListboxProvider } from './ListboxContext';
+import type { ListboxOption } from './ListboxContext';
+import type { ListboxValue } from './ListboxOption';
+import useSharedRef from '../../utils/useSharedRef';
+
+const keys = ['ArrowUp', 'ArrowDown', 'Home', 'End', 'Enter', ' '];
+
+interface ListboxProps
+  extends Omit<React.HTMLAttributes<HTMLElement>, 'onSelect'> {
+  as?: React.ElementType | string;
+  value?: ListboxValue;
+  navigation?: 'cycle' | 'bound';
+  onSelect?: <T extends HTMLElement = HTMLElement>({
+    target,
+    value
+  }: {
+    target: T;
+    value: ListboxValue;
+  }) => void;
+  onSelectionChange?: ({
+    value
+  }: {
+    previousValue: ListboxValue;
+    value: ListboxValue;
+  }) => void;
+  onActiveChange?: (option: ListboxOption) => void;
+}
+
+// id for listbox options should always be defined since it should
+// be provide via the author, or auto-generated via the component
+const getOptionId = (option: ListboxOption): string =>
+  option.element.getAttribute('id') as string;
+
+const isDisabledOption = (option: ListboxOption): boolean =>
+  option.element.getAttribute('aria-disabled') === 'true';
+
+const optionMatchesValue = (option: ListboxOption, value: unknown): boolean =>
+  typeof option.value !== null &&
+  typeof option.value !== 'undefined' &&
+  option.value === value;
+
+const Listbox = forwardRef<HTMLElement, ListboxProps>(
+  (
+    {
+      as: Component = 'ul',
+      children,
+      defaultValue,
+      value,
+      navigation = 'bound',
+      onKeyDown,
+      onFocus,
+      onSelect,
+      onSelectionChange,
+      onActiveChange,
+      ...props
+    },
+    ref
+  ): JSX.Element => {
+    const [options, setOptions] = useState<ListboxOption[]>([]);
+    const [activeOption, setActiveOption] = useState<ListboxOption | null>(
+      null
+    );
+    const [selectedOption, setSelectedOption] = useState<ListboxOption | null>(
+      null
+    );
+    const listboxRef = useSharedRef<HTMLElement>(ref);
+    const isControlled = typeof value !== 'undefined';
+
+    useLayoutEffect(() => {
+      const listboxValue = isControlled ? value : defaultValue;
+      const selectedOption = options.find((option) =>
+        optionMatchesValue(option, listboxValue)
+      );
+      setSelectedOption(selectedOption || null);
+      setActiveOption(selectedOption || null);
+    }, [options, value]);
+
+    useEffect(() => {
+      if (activeOption) {
+        onActiveChange?.(activeOption);
+      }
+    }, [activeOption]);
+
+    const handleSelect = useCallback(
+      (option: ListboxOption) => {
+        setActiveOption(option);
+        onSelect?.({ target: option.element, value: option.value as string });
+        // istanbul ignore else
+        if (!isControlled) {
+          setSelectedOption(option);
+          onSelectionChange?.({
```

Comment:
It seems that `onSelectionChange` only applies to uncontrolled `listbox`. It makes sense, cause controlled `listbox` has `onSelect`, and uncontrolled `listbox` has to rely on `onSelectionChange` if the user wants to do something whenever the selection changes. it could be confusing if a controlled listbox user accidentally passes in `onSelectionChange` and it does not do anything.

Reply:
`onSelect` happens before the selection happens. `onSelectionChange` happens only after the the change has actually happened.

I suppose we could simplify this by only having a single event that responds when a selection changes, but only actually set the selection when uncontrolled:

```tsx
const handleSelectionChange = useCallback(
  (option: ListboxOption) => {
    onSelectionChange?.({ 
      target: option.element, 
      value: option.value as string 
      previousValue: selectedOption?.value
    });

    if (!isControlled) {
      setActiveOption(option);
      setSelectedOption(option);
    }
  },
  [isControlled, selectedOption]
);
```

Would that help make it more clear?

Reply:
Personally, I like the suggested alternative. I do feel like it would make it clearer

Reply:
I just pushed that change.